File: apps/api/src/smart_badge_api/schemas/customers.py
from datetime import date, datetime

from pydantic import BaseModel

from smart_badge_api.schemas.recordings import PendingArchiveRecordingOut


class CustomerCreate(BaseModel):
    name: str
    gender: str | None = None
    age: int | None = None
    wechat_external_uid: str | None = None
    source: str | None = None
    notes: str | None = None


class CustomerUpdate(BaseModel):
    name: str | None = None
    gender: str | None = None
    age: int | None = None
    wechat_external_uid: str | None = None
    source: str | None = None
    notes: str | None = None
    is_active: bool | None = None


class CustomerOut(BaseModel):
    id: str
    name: str
    gender: str | None
    age: int | None
    wechat_external_uid: str | None
    external_customer_code: str | None = None
    source: str | None
    notes: str | None
    is_active: bool
    visit_count: int = 0
    recording_count: int = 0
    closed_won_count: int = 0
    total_deposit_principal: float | None = None
    customer_type_code: str | None = None
    customer_type_label: str | None = None
    customer_type_institution_code: str | None = None
    last_visit_at: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class CustomerDateSummaryOut(BaseModel):
    date: str | None = None
    total: int = 0


class CustomerPageOut(BaseModel):
    items: list[CustomerOut]
    total: int
    page: int
    page_size: int
    pages: int
    date_summaries: list[CustomerDateSummaryOut] = []


class CustomerDetailRecordingEvalDimensionOut(BaseModel):
    name: str
    point_score: float | None = None
    max_score: float = 1.0
    summary: str | None = None


class CustomerDetailRecordingOut(BaseModel):
    id: str
    visit_id: str | None
    file_name: str
    device_id: str | None
    staff_name: str | None = None
    status: str
    duration_seconds: int | None = None
    created_at: datetime
    transcript_id: str | None = None
    transcript_status: str | None = None
    transcript_provider: str | None = None
    transcript_excerpt: str | None = None
    analysis_task_id: str | None = None
    analysis_status: str | None = None
    analysis_overall_score: float | None = None
    analysis_completed_at: str | None = None
    analysis_summary: str | None = None
    analysis_profile_tags: list[str] = []
    analysis_primary_demands: list[str] = []
    analysis_concerns: list[str] = []
    analysis_recommendations: list[str] = []
    analysis_evaluation_dimensions: list[CustomerDetailRecordingEvalDimensionOut] = []


class CustomerDetailVisitOrderLineItemOut(BaseModel):
    fzdh: str | None = None
    advxc_long: str | None = None
    assxc: str | None = None
    fzsj: str | None = None
    fzsta_txt: str | None = None
    jcsta_txt: str | None = None


class CustomerDetailVisitOrderSummaryOut(BaseModel):
    dzdh: str | None = None
    jgbm: str | None = None
    crtdt: str | None = None
    crttm: str | None = None
    dzsta_txt: str | None = None
    dzly_txt: str | None = None
    dymd_txt: str | None = None
    dztyp_txt: str | None = None
    jgks_txt: str | None = None
    fzuer_long: str | None = None
    vipkf: str | None = None
    kulvl_dq: str | None = None
    kutyp_dq_txt: str | None = None
    kut30_dq_txt: str | None = None
    kusta_dq_txt: str | None = None
    remark_dz: str | None = None
    line_items: list[CustomerDetailVisitOrderLineItemOut] = []


class CustomerDetailVisitOut(BaseModel):
    id: str
    status: str
    external_visit_order_no: str | None = None
    visit_date: date | None = None
    visit_time: str | None = None
    consultant_name: str | None = None
    doctor_name: str | None = None
    deal_status: str | None = None
    deposit_principal: float | None = None
    deposit_bonus: float | None = None
    arrival_purpose: str | None = None
    project_needs: str | None = None
    notes: str | None = None
    created_at: datetime
    recordings: list[CustomerDetailRecordingOut]
    pending_archive_recordings: list[PendingArchiveRecordingOut] = []
    sap_consultation_texts: list[str] = []
    visit_order_summary: CustomerDetailVisitOrderSummaryOut | None = None


class CustomerDetailOut(CustomerOut):
    recording_count: int = 0
    transcript_count: int = 0
    analyzed_recording_count: int = 0
    visits: list[CustomerDetailVisitOut]


class CustomerMergedThemeOut(BaseModel):
    label: str
    detail: str | None = None
    count: int = 0
    latest_seen_at: str | None = None


class CustomerMergedDimensionOut(BaseModel):
    name: str
    average_score: float
    latest_score: float | None = None
    mention_count: int = 0
    latest_comment: str | None = None


class CustomerMergedTimelineOut(BaseModel):
    task_id: str
    recording_id: str | None = None
    recording_name: str | None = None
    visit_id: str | None = None
    visit_status: str | None = None
    project_name: str | None = None
    deal_amount: float | None = None
    overall_score: float | None = None
    quality_label: str
    completed_at: str | None = None


class CustomerMergedAnalysisOut(BaseModel):
    customer_id: str
    customer_name: str
    total_visits: int = 0
    total_recordings: int = 0
    analyzed_recordings: int = 0
    average_score: float | None = None
    latest_score: float | None = None
    score_delta: float | None = None
    score_trend: str = "stable"
    merged_summary: str = ""
    latest_task_id: str | None = None
    latest_recording_id: str | None = None
    last_analyzed_at: str | None = None
    recurring_focus_areas: list[CustomerMergedThemeOut] = []
    recurring_concerns: list[CustomerMergedThemeOut] = []
    profile_tags: list[CustomerMergedThemeOut] = []
    dimension_averages: list[CustomerMergedDimensionOut] = []
    timeline: list[CustomerMergedTimelineOut] = []


# ── 标签完成度 ──────────────────────────────────────


class TagExtractionItem(BaseModel):
    category_id: str
    category_name: str
    weight_level: int | None = None
    available_tags: list[str] = []
    extracted_values: list[str] = []
    evidence: str | None = None
    status: str = "not_extracted"  # extracted | not_extracted
    last_seen_at: str | None = None


class TagCompletionOut(BaseModel):
    customer_id: str
    total_categories: int = 0
    extracted_categories: int = 0
    completion_rate: float = 0.0
    categories: list[TagExtractionItem] = []


# ── 到诊单分组 ──────────────────────────────────────


class VisitOrderItemOut(BaseModel):
    dzseg: str | None = None
    jcsta_txt: str | None = None
    remark_dz: str | None = None


class VisitOrderGroupOut(BaseModel):
    dzdh: str
    visit_date: str | None = None
    consultant_name: str | None = None
    status_text: str | None = None
    customer_type: str | None = None
    customer_type_t30: str | None = None
    member_level: str | None = None
    remark: str | None = None
    items: list[VisitOrderItemOut] = []


class CustomerVisitOrdersOut(BaseModel):
    customer_id: str
    customer_code: str | None = None
    total_visits: int = 0
    visit_groups: list[VisitOrderGroupOut] = []

File: apps/web/src/api/customers.ts
import { api, type PaginatedResponse } from './client'

export type Customer = {
  id: string
  name: string
  gender: string | null
  age: number | null
  wechat_external_uid: string | null
  external_customer_code: string | null
  source: string | null
  notes: string | null
  is_active: boolean
  visit_count: number
  recording_count: number
  closed_won_count: number
  total_deposit_principal: number | null
  customer_type_code: string | null
  customer_type_label: string | null
  customer_type_institution_code: string | null
  last_visit_at: string | null
  created_at: string
}

export type CustomerDateSummary = {
  date: string | null
  total: number
}

export type CustomerPage = PaginatedResponse<Customer> & {
  date_summaries?: CustomerDateSummary[]
}

export type CustomerDetailRecording = {
  id: string
  visit_id: string | null
  file_name: string
  device_id: string | null
  staff_name: string | null
  status: string
  duration_seconds: number | null
  created_at: string
  transcript_id: string | null
  transcript_status: string | null
  transcript_provider: string | null
  transcript_excerpt: string | null
  analysis_task_id: string | null
  analysis_status: string | null
  analysis_overall_score: number | null
  analysis_completed_at: string | null
  analysis_summary: string | null
  analysis_profile_tags: string[]
  analysis_primary_demands: string[]
  analysis_concerns: string[]
  analysis_recommendations: string[]
  analysis_evaluation_dimensions: {
    name: string
    point_score: number | null
    max_score: number
    summary: string | null
  }[]
}

export type PendingArchiveRecording = {
  id: string
  display_file_name: string
  create_time: string | null
  duration_seconds: number | null
  staff_id: string | null
  staff_name: string | null
  device_code: string | null
  pipeline_status: string | null
  recording_id: string | null
  has_transcript: boolean
  has_analysis: boolean
  match_score: number
  match_reasons: string[]
}

export type CustomerDetailVisit = {
  id: string
  status: string
  external_visit_order_no: string | null
  visit_date: string | null
  visit_time: string | null
  consultant_name: string | null
  doctor_name: string | null
  deal_status: string | null
  deposit_principal: number | null
  deposit_bonus: number | null
  arrival_purpose: string | null
  project_needs: string | null
  notes: string | null
  created_at: string
  recordings: CustomerDetailRecording[]
  pending_archive_recordings: PendingArchiveRecording[]
  sap_consultation_texts: string[]
  visit_order_summary: {
    dzdh: string | null
    jgbm: string | null
    crtdt: string | null
    crttm: string | null
    dzsta_txt: string | null
    dzly_txt: string | null
    dymd_txt: string | null
    dztyp_txt: string | null
    jgks_txt: string | null
    fzuer_long: string | null
    vipkf: string | null
    kulvl_dq: string | null
    kutyp_dq_txt: string | null
    kut30_dq_txt: string | null
    kusta_dq_txt: string | null
    remark_dz: string | null
    line_items: Array<{
      fzdh: string | null
      advxc_long: string | null
      assxc: string | null
      fzsj: string | null
      fzsta_txt: string | null
      jcsta_txt: string | null
    }>
  } | null
}

export type CustomerDetail = Customer & {
  recording_count: number
  transcript_count: number
  analyzed_recording_count: number
  visits: CustomerDetailVisit[]
}

export type CustomerMergedTheme = {
  label: string
  detail: string | null
  count: number
  latest_seen_at: string | null
}

export type CustomerMergedDimension = {
  name: string
  average_score: number
  latest_score: number | null
  mention_count: number
  latest_comment: string | null
}

export type CustomerMergedTimelineItem = {
  task_id: string
  recording_id: string | null
  recording_name: string | null
  visit_id: string | null
  visit_status: string | null
  project_name: string | null
  deal_amount: number | null
  overall_score: number | null
  quality_label: string
  completed_at: string | null
}

export type CustomerMergedAnalysis = {
  customer_id: string
  customer_name: string
  total_visits: number
  total_recordings: number
  analyzed_recordings: number
  average_score: number | null
  latest_score: number | null
  score_delta: number | null
  score_trend: 'improving' | 'declining' | 'stable'
  merged_summary: string
  latest_task_id: string | null
  latest_recording_id: string | null
  last_analyzed_at: string | null
  recurring_focus_areas: CustomerMergedTheme[]
  recurring_concerns: CustomerMergedTheme[]
  profile_tags: CustomerMergedTheme[]
  dimension_averages: CustomerMergedDimension[]
  timeline: CustomerMergedTimelineItem[]
}

export const fetchCustomers = (params?: {
  keyword?: string
  hospital_code?: string
  is_active?: boolean
  consultant_id?: string
  has_visits?: boolean
  has_recordings?: boolean
  has_positive_recharge?: boolean
  date_from?: string
  date_to?: string
  include_date_summaries?: boolean
  fast_page?: boolean
  page?: number
  page_size?: number
}) => {
  const sp = new URLSearchParams()
  if (params?.keyword) sp.set('keyword', params.keyword)
  if (params?.hospital_code) sp.set('hospital_code', params.hospital_code)
  if (params?.is_active !== undefined) sp.set('is_active', String(params.is_active))
  if (params?.consultant_id) sp.set('consultant_id', params.consultant_id)
  if (params?.has_visits !== undefined) sp.set('has_visits', String(params.has_visits))
  if (params?.has_recordings !== undefined) sp.set('has_recordings', String(params.has_recordings))
  if (params?.has_positive_recharge !== undefined) sp.set('has_positive_recharge', String(params.has_positive_recharge))
  if (params?.date_from) sp.set('date_from', params.date_from)
  if (params?.date_to) sp.set('date_to', params.date_to)
  if (params?.include_date_summaries !== undefined) sp.set('include_date_summaries', String(params.include_date_summaries))
  if (params?.fast_page !== undefined) sp.set('fast_page', String(params.fast_page))
  if (params?.page) sp.set('page', String(params.page))
  if (params?.page_size) sp.set('page_size', String(params.page_size))
  const qs = sp.toString()
  return api.get(`customers${qs ? `?${qs}` : ''}`).json<CustomerPage>()
}

export const fetchCustomer = (id: string) => api.get(`customers/${id}`).json<Customer>()

export const fetchCustomerDetail = (id: string) =>
  api.get(`customers/${id}/detail`).json<CustomerDetail>()

export const fetchCustomerMergedAnalysis = (id: string) =>
  api.get(`customers/${id}/merged-analysis`).json<CustomerMergedAnalysis>()

export const createCustomer = (
  data: Omit<
    Customer,
    | 'id'
    | 'is_active'
    | 'visit_count'
    | 'recording_count'
    | 'closed_won_count'
    | 'total_deposit_principal'
    | 'customer_type_code'
    | 'customer_type_label'
    | 'customer_type_institution_code'
    | 'last_visit_at'
    | 'created_at'
  >,
) => api.post('customers', { json: data }).json<Customer>()

export const updateCustomer = (id: string, data: Partial<Customer>) =>
  api.put(`customers/${id}`, { json: data }).json<Customer>()

export const deleteCustomer = (id: string) => api.delete(`customers/${id}`)

// ── 标签完成度 ──────────────────────────────────

export type TagExtractionItem = {
  category_id: string
  category_name: string
  weight_level: number | null
  available_tags: string[]
  extracted_values: string[]
  evidence: string | null
  status: 'extracted' | 'not_extracted'
  last_seen_at: string | null
}

export type TagCompletion = {
  customer_id: string
  total_categories: number
  extracted_categories: number
  completion_rate: number
  categories: TagExtractionItem[]
}

export const fetchCustomerTagCompletion = (id: string) =>
  api.get(`customers/${id}/tag-completion`).json<TagCompletion>()

// ── 到诊单历史 ──────────────────────────────────

export type VisitOrderItem = {
  dzseg: string | null
  jcsta_txt: string | null
  remark_dz: string | null
}

export type VisitOrderGroup = {
  dzdh: string
  visit_date: string | null
  consultant_name: string | null
  status_text: string | null
  customer_type: string | null
  customer_type_t30: string | null
  member_level: string | null
  remark: string | null
  items: VisitOrderItem[]
}

export type CustomerVisitOrders = {
  customer_id: string
  customer_code: string | null
  total_visits: number
  visit_groups: VisitOrderGroup[]
}

export const fetchCustomerVisitOrders = (id: string) =>
  api.get(`customers/${id}/visit-orders`).json<CustomerVisitOrders>()
